File: Graph/dfs_string/word_break.kt
import java.util.*


// This is actually a bfs solution here
/*

 s = "applepenapple",
  wordDict = ["apple","pen"]
so the word s can be separated here
 */

fun main() {
    var word = "leetcode"
    var wordDict= listOf("leet", "code")
    wordBreak(word, wordDict)
}
    // from a node start we can move to end if
    // substrig s between start, end exists in wordDict

fun wordBreak(s: String, wordDict: List<String>?): Boolean {

   // Uique qoreds
    val words: Set<String> = HashSet(wordDict)
    val queue: Queue<Int> = LinkedList()

    // Store the index of the array seen
    val seen = BooleanArray(s.length + 1)
    queue.add(0)
    while (!queue.isEmpty()) {
        /*
Using exmample baove, the first time it comes here the
start will then be 4 the next time is 8
         */
        val start = queue.remove()
        println("start is $start")
        if (start == s.length) {
            return true
        }


        for (end in start + 1..s.length) {
            if (seen[end]) {
                continue
            }
            // say the word is
            if (words.contains(s.substring(start, end))) {
                queue.add(end)
                seen[end] = true
            }
        }
    }
    return false
}

File: Graph_string/word_break.py
'''


Intuition

Let's imagine the indices of s as a graph. Each index can be thought of as a node, which represents building s up to the index.

Adding a word to an existing string is like an edge between nodes. '
For a node start, we can move to node end if the substring of s between start, end exists in wordDict.



For example, let's say we have s = "leetcode" and wordDict = ["leet", "code"].

We are currently at node 4, which implies that we have built "leet" (the first 4 characters of s).
We can move to node 8, because the substring of s with indices [4, 8) is "code", which is in wordDict.


We start at node 0, which represents the empty string. We want to reach node s.length, which implies that we have
 built the entire string. We can run a BFS to accomplish this traversal.

  If you're not familiar with BFS, check out the relevant Explore Card.

At each node start, we iterate over all the nodes end that come after start.

For each end, we check if the substring between start, end is in wordDict. If it is, we can add end to the queue.

    We will first convert wordDict into a set so that we can perform the checks in constant time. We will also use a data structure seen to prevent us from visiting a node more than once.
'''
from collections import deque


def wordBreak(self, s: str, wordDict: List[str]) -> bool:
    words = set(wordDict)
    queue = deque([0])
    seen = set()

    while queue:
        start = queue.popleft()
        if start == len(s):
            return True

        for end in range(start + 1, len(s) + 1):
            if end in seen:
                continue

            if s[start:end] in words:
                queue.append(end)
                seen.add(end)

    return False